source/extensions/filters/common/local_ratelimit/local_ratelimit_impl.cc

```diff
@@ -13,16 +13,76 @@ namespace Filters {
 namespace Common {
 namespace LocalRateLimit {
 
+SINGLETON_MANAGER_REGISTRATION(local_ratelimit_share_provider_manager);
+
+class DefaultEvenShareMonitor : public ShareProviderManager::ShareMonitor {
+public:
+  uint32_t tokensPerFill(uint32_t origin_tokens_per_fill) const override {
+    return std::ceil(origin_tokens_per_fill * share_);
+  }
+  void onLocalClusterUpdate(const Upstream::Cluster& cluster) override {
+    const auto number = cluster.info()->endpointStats().membership_total_.value();
+    share_ = number == 0 ? 1.0 : 1.0 / number;
+  }
+
+private:
+  double share_{1.0};
```

Comment:
Can this data member be accessed by multiple threads?

Reply:
This is only accessed in the main thread because the token filling timer is executed in the main thread.

Reply:
drive-by: assertions that code is only running on the main thread can easily be proved in the code. See https://github.com/envoyproxy/envoy/blob/81e37b0ccd99934189f2a018c63dc6a654b1afe2/source/common/common/thread.h#L300

Reply:
I agree that using ASSERT_IS_MAIN_OR_TEST_THREAD should be used here, and it will clarify which thread access this data member.

Reply:
Done. Thanks.